packages/script/src/script-factory.ts

```diff
@@ -0,0 +1,106 @@
+import type { BytesLike } from '@ethersproject/bytes';
+import { Logger } from '@ethersproject/logger';
+import { Interface } from '@fuel-ts/abi-coder';
+import type { JsonAbi } from '@fuel-ts/abi-coder';
+import type { TransactionResponse, TransactionResult } from '@fuel-ts/providers';
+import { getDecodedLogs } from '@fuel-ts/providers';
+import { ReceiptType } from '@fuel-ts/transactions';
+import { versions } from '@fuel-ts/versions';
+import type { BaseWalletLocked } from '@fuel-ts/wallet';
+
+import callScript from './callScript';
+import { Script } from './script';
+
+const logger = new Logger(versions.FUELS);
+
+const FUNCTION_FRAGMENT_NAME = 'main';
+
+type Data = any[];
+
+type Result = {
+  value: any;
+  logs: any[];
+};
+
+export default class ScriptFactory {
+  bytecode: BytesLike;
+  script: Script<any, any>;
+  interface: Interface;
+  wallet: BaseWalletLocked;
+
+  constructor(bytecode: BytesLike, abi: JsonAbi | Interface, wallet: BaseWalletLocked) {
+    this.bytecode = bytecode;
+    this.wallet = wallet;
+    if (abi instanceof Interface) {
+      this.interface = abi;
+    } else {
+      this.interface = new Interface(abi);
+    }
```

Comment:
Overloading the `abi` parameter seems unnecessary; we could accept `JsonAbi` only.

```suggestion
  constructor(bytecode: BytesLike, abi: JsonAbi, wallet: BaseWalletLocked) {
    this.bytecode = bytecode;
    this.wallet = wallet;
    this.interface = new Interface(abi);
```